src/server/server.rs

```diff
@@ -204,13 +206,23 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver> Server<T, S> {
         let msg_type = msg.get_msg_type();
         match msg_type {
             MessageType::Raft => {
+                let msg_type = msg.get_raft().get_message().get_msg_type();
                 if let Err(e) = self.raft_router.rl().send_raft_msg(msg.take_raft()) {
                     // Should we return error to let outer close this connection later?
                     error!("send raft message for token {:?} with msg id {} err {:?}",
                            token,
                            msg_id,
                            e);
                 }
+                if msg_type == RaftMessageType::MsgSnapshot {
+                    // Now we use new connection for every snapshot sending,
+                    // but it is not safe to close the connection after sending OK,
+                    // so we close it here after receiving whole snapshot.
+                    // We use shutdown here for a graceful close.
```

Comment:
tcp socket guarantee it. 
After shutdown, the server will send the client FIN, the client knows all data sent to server successfully, then close the socket. Then the server will read 0 later and knows the socket is closed.